### ISSUE TITLE
Crash when restoring saved game

### ISSUE THREAD

**Issue by Crystal-Shard** [adventuregamestudio/ags#2073]
**Describe the bug**
Saved games created under an earlier engine version crash when opened in 3.6. The error message is as follows:
```
An internal error has occurred. Please note down the following information.
If the problem persists, contact the game author for support or post these details on the AGS Technical Forum.
(Engine version 3.6.0.47)

Error: Can't deserialize interaction: too many events
```

**AGS Version**
Engine version 3.6.0.47

**Game**
Heroine's Quest

**To Reproduce**
1. Open a saved game that was created in an earlier engine version. I can provide executable and saved game(s) if it helps.

**Expected behavior**
I'd expect saved games to be compatible (assuming the base game hasn't changed).

**Desktop (please complete the following information):**
Windows 11

**Comment by ivan-mogilko** [comment]
> I can provide executable and saved game(s) if it helps.

Yes please, it's better to have a actual save and point to exact game version, in case this has a specific condition to replicate.

**Comment by Crystal-Shard** [comment]
Please check https://crystalshard.net/HQ3.6.zip
The saved game goes in %homepath%\Saved Games\Heroine's Quest 1.2

**Comment by ivan-mogilko** [comment]
There's something very wrong about this. The "interactions" are legacy component that is not met in the game data since after AGS 2.72.

Another weird thing is that the save is loaded fine if you run the original HQ1.2 (made in 3.3.0) using the 3.6.0 engine.

**Comment by Crystal-Shard** [comment]
Interesting. Is that the old "interaction editor" that was basically a low-code language for people who didn't want C-style scripting? Because I'm very sure I've never used that.

I note the old acsetup.cfg file for HQ contains "letterbox=0" and "forceletterbox=1", I do not recall why that is there. The game runs in 320x200 (or a multiple thereof) so it should not be letterboxed.

**Comment by ivan-mogilko** [comment]
> The game runs in 320x200 (or a multiple thereof) so it should not be letterboxed.

The old AGS had a quirk where the game resolution would be set to 320x200, and the option "letterbox" was set in General Settings, which in combination produced 320x240 game with a game screen centered vertically if the room is smaller (like 320x200). The option "forceletterbox" probably overrides this, but the reason why it was not applied in original game is not fully clear to me, and frankly I am not interested enough to investigate this.

**Comment by Crystal-Shard** [comment]
Right. At least for the games I'm aware of, ignoring this sounds like the best approach.

**Comment by ivan-mogilko** [comment]
I looked into why the saves cannot be loaded, and that's because of my mistake that I introduced around early development of AGS 3.4.0. There has been a change to save data, but instead of rising save format index, and checking for that, I checked for the version of the *current game*.

This means that if the game is upgraded, the engine won't be able to load the older saves anymore.

I looked around a bit, but cannot see any option to let engine properly detect that the save was made prior to 3.4.0 if the current game is 3.4.0 or higher.

**Comment by ivan-mogilko** [comment]
> I looked around a bit, but cannot see any option to let engine properly distinguish that the save was made prior to 3.4.0 if the current game is 3.4.0 or higher. The save data simply does not contain anything that would indicate that.

There might be a way to deduce that the save does not belong to 3.4.0 or higher, by analyzing some of the game data, but that seem to be not 100% reliable, because of multiple mistakes I did when changing save format.

@Crystal-Shard please tell, how important this is to you?

We have a completely new save format starting with AGS 3.5.0, and the previous save "generation" was kept supported temporarily, but as the time goes this seem to be rather redundant, and I've been considering to remove this support in AGS 3.6.1.

**Comment by Crystal-Shard** [comment]
Alternatively, one of the options in the AGS editor is "script compatibility level" so perhaps there could be a pulldown for savegame compatibility level? Because generally speaking I'd expect each individual game to have only one kind of saved game.

I'd say it's a pretty big deal for games on Steam (and similar sites) to not break backwards compatibility. That is, Steam usually forces updates, so if a game updates and this invalidates all existing saved games for everyone, that would upset a lot of players. Conversely, if an update was opt-in only, then I'd expect most players not to notice and not to update. I'm curious how other full-length games (e.g. anything by Dave Gilbert) handle this?

I suppose a viable alternative for me is to _not_ rebuild the game, but simply replace the AGS executable by the 3.6 one.

**Comment by ivan-mogilko** [comment]
> Because generally speaking I'd expect each individual game to have only one kind of saved game.
> I'd say it's a pretty big deal for games on Steam (and similar sites) to not break backwards compatibility. That is, Steam usually forces updates, so if a game updates and this invalidates all existing saved games for everyone, that would upset a lot of players.

I wish this was discussed *years* ago, when the open source project started, and explicitly defined as a part of the engine's purpose and design; but this was never done. In these years I kept a goal to support running old games, but loading saves from old versions was seen as rather as a bonus, which helped to do a comparison tests between game runs.

That resulted in me being more careless about save format, as I had impression that it has to be kept valid only for a few consequent versions. Indeed I should have known better, but unfortunately there were no-one to oversee my work at that early time.

In regards to old support, today we even decided to remove the whole compatibility layer in the development of a next major release (ags 4.0). This removes both support of saved games, but also a whole bunch of script commands.

---

> I suppose a viable alternative for me is to not rebuild the game, but simply replace the AGS executable by the 3.6 one.

This should work.

> Alternatively, one of the options in the AGS editor is "script compatibility level" so perhaps there could be a pulldown for savegame compatibility level?

This is theoretically possible, assuming the engine maintains at least a basic ability to read old formats. But it's not clear how this should be used. Old saves format cannot have new features saved in them. If you will use them in your game, these will not be written to saves? Also this adds to the number of features that I'll have to support and maintain. Should the same thing be kept maintained in the "ags 4" version? This is not something to push into the engine randomly...

**Comment by ivan-mogilko** [comment]
> I will look more into this particular format problem, like mentioned, there may be a way to workaround it with some ugly hack, which may work for now, but I do not have a guarantee...

No, unfortunately, this is not possible. I thought about using some game option values to detect this, but the old save format is inconvenient in a sense that these are written too late, much later than the reading error that takes place in this case.

The remaining solutions I see here are:
1) Don't upgrade the game data, as you proposed, and only replace the engine executable. The engine will simply run the old game data, as it's meant to support this. This is most straightforward, although that cuts you off from any later game features.
2) Introduce a hack, either as a game option, or a config option, that would instruct the engine to treat old saves *as if they were saved by an engine of particular version*.

**Comment by ivan-mogilko** [comment]
@Crystal-Shard ,

so here's the minimal experimental version, based on latest 3.6.0;
code branch: https://github.com/ivan-mogilko/ags-refactoring/tree/360--pretend-gamever-for-old-saves
engine exe for windows: https://www.dropbox.com/s/eiedrvmpowp4py1/360-experiment-oldsavsfix.zip?dl=0

This build supports a new config setting that tells the engine which assumptions to make about older engine version which made these saves. The setting is done like this:
```
[override]
dataver_for_legacysaves=43
```
The number is taken from the list of data formats, here:
https://github.com/adventuregamestudio/ags/blob/master/Common/ac/game_version.h
If not set, then the current game's data version is used.
But this is used **only** *when loading old saves*; the engine will ignore this setting when loading new saves, because they already have necessary information in their header.

With this I was able to load this save in *both* old and new variants of "Heroine Quest".

If this is something that you may use, we may discuss further improvement, like the way the setting's value is formed, and so forth,

**Comment by Crystal-Shard** [comment]
@ivan-mogilko Thank you for your quick response!

I've tested your build and it works for me as well. I believe this is a good approach and would appreciate using this. It's entirely fine if saved games from the new 3.6 engine no longer work under 3.3 (I'd find it highly unlikely for any player to want that); and from my own version control I can see that every published build of this game uses the same AGS version (i.e. 3.3.0).

Although I have no strong opinion on where to put the config setting, I'd say if would be most straightforward to use game_start() to set a variable like Game.LegacySaveVersion = 43;

**Comment by ivan-mogilko** [comment]
I'm sorry, I got distracted by other things recently, but now I am back at this. Do I understand correctly that you need this in 3.6.0? If so I would have to release this as a new 3.6.0 patch (otherwise I could add it in 3.6.1, but 3.6.1 is still in beta development stage).

> Although I have no strong opinion on where to put the config setting, I'd say if would be most straightforward to use game_start() to set a variable like Game.LegacySaveVersion = 43;

I would not like to add an *explicit* property in script API for this, because this solution is essentially a hack meant to workaround a mistake made in the older engine version, useful only in a rather narrow case. In addition to that, this relies on engine keep supporting very old save format. Therefore I don't want to "solidify" this in API.
For a while I've been considering adding this to extendable "game options" (ones that are set using SetGameOption function), but am still in doubts for the similar reasons... I will think a little more. But the "override" in config seem to be most flexible at the moment, as it does not add any long-term responsibilities to the engine... In theory, it's still possible to add a game option later, if the config method would be found unsuitable.

**Comment by Crystal-Shard** [comment]
No worries. Yes, I would appreciate having this in the stable (3.6.0) version. And fair point about not having this in the API; it is entirely workable for me to put it in the config file instead.

**Comment by ivan-mogilko** [comment]
I pushed these changes to our release-3.6.0 branch. They will be included in the next 3.6.0 patch, supposedly released in the beginning of September.

If you like, there's a temp build found on our build server:
https://cirrus-ci.com/task/5872449821081600

**Comment by Crystal-Shard** [comment]
Thank you, I've been testing this and so far it works fine.

The downside of this approach is that building from the AGS editor overwrites the config file (thus removing the [override] block) so I have to manually re-add it each time. That's not a big deal though.

I note that your new build fixes the earlier issue I had with windows security (i.e. windows security would delete the game executable, unless I use the setting to attach the .ags data file to the executable).

However: loading older saved games **will** crash if I do _not_ attach the .ags data file to the executable. If I recall correctly, saved games contain the name of the executable (and will terminate if the .exe file is renamed); so possibly in this case the saved game contains "Heroine's Quest.exe" and the (non-attached) game program is called "Heroine's Quest.ags".

**Comment by ivan-mogilko** [comment]
> The downside of this approach is that building from the AGS editor overwrites the config file (thus removing the [override] block) so I have to manually re-add it each time. That's not a big deal though.

Right, I forgot about this. But there's a solution since 3.6.0, as noted in a changelog: Editor now copies a acsetup.cfg file from the project's root folder if one is provided by a user, as a base for generating default config file. Settings from Default Setup will be written over, but custom options won't be touched.

> loading older saved games will crash if I do not attach the .ags data file to the executable. If I recall correctly, saved games contain the name of the executable (and will terminate if the .exe file is renamed); so possibly in this case the saved game contains "Heroine's Quest.exe" and the (non-attached) game program is called "Heroine's Quest.ags".

I wonder if this is because it tries to load game data from "Heroine's Quest.exe" not realizing that it does not contain one. This has to be fixed.

**Comment by Crystal-Shard** [comment]
> I wonder if this is because it tries to load game data from "Heroine's Quest.exe" not realizing that it does not contain one. This has to be fixed. EDIT: hm, no, this does not reproduce in a simple test case. Could you tell which error do you get exactly? Or is there a way to get this variant of your compiled game for test?

I don't get an error, the game simply crashes. You can try in this build http://crystalshard.net/test/crash.zip (loading the saved game in the zipfile).

I've found another thing: if I restore a saved game where the player has selected another language, then the game will revert to English again. Earlier behavior was to use the language used when saving the game.

**Comment by ivan-mogilko** [comment]
> if I restore a saved game where the player has selected another language, then the game will revert to English again. Earlier behavior was to use the language used when saving the game.

This may have a number of factors; how is language set in your game, is it remembered in some variable?

**Comment by Crystal-Shard** [comment]
The game opens with a screen that shows all languages (because this is more visible to players than in the setup tool), then uses Game.ChangeTranslation() based on that.
Later in the game, Game.TranslationFilename is used in a few places where GUIs or fonts need to be changed based on the language.

________________________________

**Comment by Crystal-Shard** [comment]
> > http://crystalshard.net/test/crash.zip
> > This link gives me 404 error...

Sorry, that should be https. https://crystalshard.net/test/crash.zip

**Comment by ivan-mogilko** [comment]
> I find that if I restore an "old" saved game, the save it again (under a new name), then load that new save, the game crashes with error:

I was not able to reproduce this with the game version that you linked earlier and this save. This save loads fine, and if I save another - that also loads fine.
From what I can see, the game has 82 gui labels, and save references same amount.

TBH, I don't really see how restoring "old" save may make any difference... Does it not happen with just any save?

**Comment by ivan-mogilko** [comment]
> I find that if I restore an "old" saved game, the save it again (under a new name), then load that new save, the game crashes with error:

So apparently you had changed the number of labels in game since the version that old saves were done in (78 to 82).
Changing number of game objects usually leads to old saves becoming incompatible.

This was not noticed because there's another logical mistake in a function that loads old saves: it does not test for mismatching GUI numbers, but instead changes that size. In other words, reading old save with less labels make some of the newer labels "disconnect" from the game.

**Comment by ivan-mogilko** [comment]
Regarding the game language issue:

> I've found another thing: if I restore a saved game where the player has selected another language, then the game will revert to English again. Earlier behavior was to use the language used when saving the game.

The translation is not a part of the save. I am not 100% certain about old AGS (I will check), but it's been like this for a while now. Imo this is logical, because translation is player's setting, not a game's data. A player may want to load somebody elses save, but use their own translation. Or replay their own game using a different translation.

I don't know how it worked before. But in theory it might be possible to save current translation in a variable, and then test that variable after restoring a save, and init corresponding translation: that is if you prefer to keep same translation when a save is loaded.

**Comment by ivan-mogilko** [comment]
I fixed this problem:

> However: loading older saved games will crash if I do not attach the .ags data file to the executable. If I recall correctly, saved games contain the name of the executable (and will terminate if the .exe file is renamed); so possibly in this case the saved game contains "Heroine's Quest.exe" and the (non-attached) game program is called "Heroine's Quest.ags".

The new build may be downloaded here:
https://cirrus-ci.com/task/6158281102589952

**Comment by Crystal-Shard** [comment]
I confirm that the new build (5720) does not crash over a non-attached .ags data file, nor over an incorrect amount of labels/buttons. Thank you for your help!

> So apparently you had changed the number of labels in game since the version that old saves were done in (78 to 82). Changing number of game objects usually leads to old saves becoming incompatible.
Yes, because I tested this rather extensively and it worked without issue. It's a good point that it causes labels to "disconnect", but (almost) all GUI labels I use are never altered at runtime, so I never noticed that this could be an issue. In my opinion, a save that may causes a visual discrepancy is preferable over a save that crashes.

> **It seems like these saves were made in different versions of HQ, and there had been a change to buttons and labels since, which went unnoticed because of the lack of checks in the engine... At least that's my only explanation.
Yes, that sounds right. I can check in my git repo if you'd like to know exactly what changed. It's probably an added help text or something. Essentially, HQ's saved games have been forward-compatible since 2014 (and players have appreciate that).

> The translation is not a part of the save. I am not 100% certain about old AGS (I will check), but it's been like this for a while now. Imo this is logical, because translation is player's setting, not a game's data.
I agree that this is a good approach. Sorry for misremembering.

**Comment by Crystal-Shard** [comment]
> But it won't work in general case, where a game may have labels accessed at runtime. Frankly, I am not certain how did that even worked for you, as if label or a button is disconnected, that means it cannot be displayed or interacted with, all of its data will be lost. Maybe players somehow did not notice, or ignored that.
I've figured out where it's coming from. The added labels and buttons are all on the title screen and title menus, and you can't save the game at that point. At any time when the game has actually started and the player can save, the title screen is no longer accessible (except via RestartGame()) so the player can never notice this glitch.

> the rest will remain in the state they've been before the save, as currently there's no mechanism to restore these even to a initial state.
So in this case, that would work (as the buttons/labels are never altered at runtime, and are never actually shown after a game is restored). I suppose setting them to blank or zero would also work.

> Please note that the **new** saves are totally strict about this, and they will prevent loading a save if it does not contain same amount of game objects as the current game.
Yes, I'll keep note for future games (Sir Bob, in particular).

> > I confirm that the new build (5720) does not crash over a non-attached .ags data file, nor over an incorrect amount of labels/buttons.
Upon further investigation, my statement is in error. As you say, at least _some_ of the older saves to no longer load.
 
> Normally I would be against this, but if keeping these old saves is a big deal, then these assertions could be ignored by some kind of "override" config option again. Of course this implies that the game will be in a unsafe state after loading such save...
I would appreciate that, because supporting both a 3.6 and a 3.3 branch of Heroine's Quest on sites like Steam is rather messy.
If I understand correctly, the issue is that an old save has (say) 78 controls, whereas the game has 82. Back in AGS3.3, loading this would result in a game state with 82 controls (the last four disconnected); and in AGS3.6 with the legacysaves override, it results in a game state with 78 controls (so saving this again will result in an invalid saved game).
So maybe there's some assignment like Game.Controlcount = SavedGame.ControlCount, and by removing that (within the context of the legacysaves override) it should work again?

**Comment by ivan-mogilko** [comment]
On a side note, in the old save format the GUI were saved exact same way as they were saved in a game data (this is different from other objects, which had some parts separated in memory). That included everything, event handlers too. This means that loading a gui from save was like loading full data, which could make it working even without syncing with the game data.

Furthermore, because of that, the saves would keep the names of registered functions for gui events.

I excluded this data when designing a new save format, because there's no way to configure these things at runtime, but now I wonder if that was done for simplicity sake, or intentional...

In regards to saves syncing problem, there's currently a planned move towards making all object types dynamic (being able to create them at runtime), like #2122. I think the situation with saves will change a lot after this.

**Comment by Crystal-Shard** [comment]
Ok, that sounds like a good solution but it's also a long-term one. In the short term, would it be possible to have the game not crash if a saved game contains too few buttons/labels? I'm fine with it if those buttons/labels disappear (as noted above, they're on the title screen so the player won't notice), and I'm also fine if this is gated behind some config option.

Basically, the website GOG has expressed interest in our game, and I would like to give them a version that both runs on the newest (non-beta) AGS engine, and also does not crash for players that have played the game earlier (and want to use their old saves). Thank you for your time!

**Comment by ivan-mogilko** [comment]
@Crystal-Shard here's a build I made for a test, it lets load legacy saves with less guis; but it reserves data before save and copies over missing items, in order to keep the numbers same as before.
https://www.dropbox.com/s/6axqnrjlgpltuuv/acwin-360-letguidiff.zip?dl=0

Please test to see if it does not cause any strange issues, then I will add this to 3.6.0 patch branch.

**Comment by Crystal-Shard** [comment]
Thank you. I can confirm that this works.

Just to check with you, it will check for the [override] section in acsetup.cfg in the game's executable folder, not in the saved game folder, is that correct? Or maybe it does both, that's also fine. The reason is that I can have Steam automatically update the setup in the executable folder (but not the other one).

**Comment by ivan-mogilko** [comment]
> Just to check with you, it will check for the [override] section in acsetup.cfg in the game's executable folder, not in the saved game folder, is that correct?

Engine checks everywhere, but having it in the default config is a right thing to do, for the obvious reasons that you stated.
Overall the configs are merged from various sources (command line too, for options that are supported via cmd line), and then engine applies this merged config.

**Comment by Crystal-Shard** [comment]
Most of the saves work fine, but a handful show something odd. When opening this save, if you move the mouse to the top of the screen, the command bar looks odd. This is the first GUI in the game, and its first buttons are being overwritten (by those of the last GUI, #49, which had those extra buttons/labels added later on). If you restore the same saved game a second time, these buttons are instead blank. I can fix this in code, but I thought I should let you know. 

![image](https://github.com/adventuregamestudio/ags/assets/107799718/e3d8e423-479c-4ea1-a896-9c11ff5ebde4)
https://crystalshard.net/test/agssave.033.hqthor

**Comment by AlanDrake** [comment]
Forcing savestates to work with different game data is fertile ground for headaches.
My opinion is that saveststes should always break to avoid undefined behaviors.

**Comment by ivan-mogilko** [comment]
@Crystal-Shard how many saves are there with the less gui? How old are these? They must be older than the current 3.3.0-based version of the game.

Alan is right, this is going to cause a lot of trouble.

I have a very bad feeling about this, and don't like when i have to put more hacks into the engine. Especially since it is related to the very old save versions.

I'd rather not put this into 3.6.0 after all.

**Comment by ivan-mogilko** [comment]
@Crystal-Shard I'm sorry, I simply made a math mistake when writing guis over. So instead the last newer guis rewrote the first guis in the list.

Here's the fixed build:
https://www.dropbox.com/s/6axqnrjlgpltuuv/acwin-360-letguidiff.zip?dl=0

**Comment by Crystal-Shard** [comment]
Works like a charm. Awesome, and thank you for your help!

**Comment by ivan-mogilko** [comment]
All the related fixes and config options are now a part of 3.6.0 Patch 5 release.

@Crystal-Shard please note that I renamed the config option names, for consistency, they are now called:
- legacysave_assume_dataver
- legacysave_let_gui_diff
respectively.
